// -----------------------------------------------------------------------------
// Copyright 2018 Frank V. Castellucci
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";
option java_multiple_files = true;
option java_package = "hashblock.events.protobuf";

// Resource quantity
message Quantity {

    //  Quantity amount? - Fixed width integer
	bytes value = 1;

    //  Address of unit of measure (e.g. bag) prime numbers
	bytes valueUnit = 2;

    //  Address of resource (e.g. peanuts) prime numbers
	bytes resourceUnit = 3;
}

// Quantity ratio
message Ratio {

    //  Transaction quantity of unit and resource 
    Quantity numerator = 1;

    //  Transaction quantity of unit and resource 
    Quantity denominator = 2;
}

// Economic event with plus/minus duality
message InitiateEvent {
    //  Keys of resource inc.
    bytes plus = 1;
    //  Keys of resource dec.
    bytes minus = 2;
    //  Transaction quantity of unit and resource 
    Quantity quantity = 3;
}

message ReciprocateEvent {
    //  Keys of transaction inc. has to jive with IE minus
    bytes plus = 1;
    //  Keys of resource dec. has to jive with IE plus
    bytes minus = 2;
    //  Don't know how this is used
    Ratio ratio = 3;
    //  Transaction quantity of unit and resource
    Quantity quantity = 4;
    //  Address of initiating event - maybe
    bytes initiate_event = 5;
}

message EventPayload {
    enum Action {
        INITIATE_EVENT = 0;
        RECIPROCATE_EVENT = 1;
    }

    //  See enum above
    Action action = 1;

    //  The event key
    string key = 2;

    // The content of this payload
    bytes data = 3;
}
